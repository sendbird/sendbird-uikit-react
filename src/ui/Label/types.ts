export const Typography = {
  H_1: 'H_1',
  H_2: 'H_2',
  SUBTITLE_1: 'SUBTITLE_1',
  SUBTITLE_2: 'SUBTITLE_2',
  BODY_1: 'BODY_1',
  BODY_2: 'BODY_2',
  BUTTON_1: 'BUTTON_1',
  BUTTON_2: 'BUTTON_2',
  BUTTON_3: 'BUTTON_3',
  CAPTION_1: 'CAPTION_1',
  CAPTION_2: 'CAPTION_2',
  CAPTION_3: 'CAPTION_3',
} as const;

export const Colors = {
  ONBACKGROUND_1: 'ONBACKGROUND_1',
  ONBACKGROUND_2: 'ONBACKGROUND_2',
  ONBACKGROUND_3: 'ONBACKGROUND_3',
  ONBACKGROUND_4: 'ONBACKGROUND_4',
  ONCONTENT_1: 'ONCONTENT_1',
  ONCONTENT_2: 'ONCONTENT_2',
  PRIMARY: 'PRIMARY',
  ERROR: 'ERROR',
  SECONDARY_3: 'SECONDARY_3',
} as const;
