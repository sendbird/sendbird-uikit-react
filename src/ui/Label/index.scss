@import '../../styles/variables';

.sendbird-label {
  -webkit-user-select: none;
  -webkit-touch-callout: none;
}

[class*=sendbird-label] {
  font-family: var(--sendbird-font-family-default);
}

.sendbird-label--h-1 {
  font-size: 20px;
  font-weight: 600;
  font-stretch: normal;
  font-style: normal;
  line-height: 1.4;
  letter-spacing: -0.2px;
}

.sendbird-label--h-2 {
  font-size: 18px;
  font-weight: 600;
  font-stretch: normal;
  font-style: normal;
  line-height: normal;
  letter-spacing: -0.2px;
}

.sendbird-label--subtitle-1 {
  font-size: 16px;
  font-weight: normal;
  font-stretch: normal;
  font-style: normal;
  line-height: 1.38;
  letter-spacing: -0.2px;
}

.sendbird-label--subtitle-2 {
  font-size: 14px;
  font-weight: 500;
  font-stretch: normal;
  font-style: normal;
  line-height: 1.14;
  letter-spacing: -0.2px;
}

.sendbird-label--body-1 {
 font-size: 14px;
 font-weight: normal;
 font-stretch: normal;
 font-style: normal;
 line-height: 1.43;
 letter-spacing: normal;
}

.sendbird-label--body-2 {
  font-size: 12px;
  font-weight: normal;
  font-stretch: normal;
  font-style: normal;
  line-height: 1.33;
  letter-spacing: normal;
}

.sendbird-label--button-1 {
  font-size: 14px;
  font-weight: 600;
  font-stretch: normal;
  font-style: normal;
  line-height: 1.43;
  letter-spacing: normal;
}

.sendbird-label--button-2 {
  font-size: 14px;
  font-weight: normal;
  font-stretch: normal;
  font-style: normal;
  line-height: 1.43;
  letter-spacing: normal;
}

.sendbird-label--caption-1 {
 font-size: 14px;
 font-weight: 600;
 font-stretch: normal;
 font-style: normal;
 line-height: 1.43;
 letter-spacing: normal;
}

.sendbird-label--caption-2 {
 font-size: 12px;
 font-weight: bold;
 font-style: normal;
 line-height: 1;
 letter-spacing: normal;
}

.sendbird-label--caption-3 {
  font-size: 12px;
  font-weight: normal;
  font-stretch: normal;
  font-style: normal;
  line-height: 1;
  letter-spacing: normal;
}

.sendbird-label--color-onbackground-1 {
  @include themed() {
    color: t(on-bg-1);
  }
}

.sendbird-label--color-onbackground-2 {
  @include themed() {
    color: t(on-bg-2);
  }
}

.sendbird-label--color-onbackground-3 {
  @include themed() {
    color: t(on-bg-3);
  }
}

.sendbird-label--color-onbackground-4 {
  @include themed() {
    color: t(on-bg-4);
  }
}

.sendbird-label--color-oncontent-1 {
  @include themed() {
    color: t(on-content-1);
  }
}

.sendbird-label--color-oncontent-2 {
  @include themed() {
    color: t(on-content-2);
  }
}

.sendbird-label--color-primary {
  @include themed() {
    color: t(primary-3);
  }
}

.sendbird-label--color-error {
  @include themed() {
    color: t(error--3-2);
  }
}

.sendbird-label--color-secondary-3 {
  @include themed() {
    color: t(secondary-3);
  }
}
