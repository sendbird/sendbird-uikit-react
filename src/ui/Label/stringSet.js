const getStringSet = (lang = 'en') => {
  const stringSet = {
    en: {
      OPEN_CHANNEL_SETTINGS__OPERATOR_TITLE: 'Channel Information',
      OPEN_CHANNEL_SETTINGS__OPERATOR_URL: 'URL',
      OPEN_CHANNEL_SETTINGS__PARTICIPANTS_ACCORDION_TITLE: 'Participants',
      OPEN_CHANNEL_SETTINGS__DELETE_CHANNEL_PANEL: 'Delete channel',
      OPEN_CHANNEL_SETTINGS__DELETE_CHANNEL_TITLE: 'Delete this channel',
      OPEN_CHANNEL_SETTINGS__DELETE_CHANNEL_SUBMIT: 'Delete',
      OPEN_CHANNEL_SETTINGS__PARTICIPANTS_TITLE: 'Participants',
      OPEN_CHANNEL_SETTINGS__EMPTY_LIST: 'No participants yet',
      OPEN_CHANNEL_SETTINGS__SEE_ALL: 'See all participants',
      OPEN_CHANNEL_SETTINGS__ALL_PARTICIPANTS_TITLE: 'All participants',
      OPEN_CHANNEL_SETTINGS__NO_TITLE: '(No title)',
      OPEN_CHANNEL_CONVERSATION__TITLE_PARTICIPANTS: 'participants',
      TRYING_TO_CONNECT: 'Trying to connectâ€¦',
      USER_PROFILE__MESSAGE: 'Message',
      USER_PROFILE__USER_ID: 'User ID',
      EDIT_PROFILE__TITLE: 'My profile',
      EDIT_PROFILE__IMAGE_LABEL: 'Profile image',
      EDIT_PROFILE__IMAGE_UPLOAD: 'Upload',
      EDIT_PROFILE__NICKNAME_LABEL: 'Nickname',
      EDIT_PROFILE__NICKNAME_PLACEHOLDER: 'Enter your nickname',
      EDIT_PROFILE__USERID_LABEL: 'User ID',
      EDIT_PROFILE__THEME_LABEL: 'Dark theme',
      CHANNEL__MESSAGE_INPUT__PLACE_HOLDER: 'Enter message',
      CHANNEL__MESSAGE_INPUT__PLACE_HOLDER__DISABLED: 'Chat is unavailable in this channel',
      CHANNEL__MESSAGE_INPUT__PLACE_HOLDER__MUTED: 'Chat is unavailable because you are being muted',
      CHANNEL__MESSAGE_LIST__NOTIFICATION__NEW_MESSAGE: 'new message(s) since',
      CHANNEL__MESSAGE_LIST__NOTIFICATION__ON: 'on',
      CHANNEL_SETTING__HEADER__TITLE: 'Channel information',
      CHANNEL_SETTING__PROFILE__EDIT: 'Edit',
      CHANNEL_SETTING__MEMBERS__TITLE: 'Members',
      CHANNEL_SETTING__MEMBERS__SEE_ALL_MEMBERS: 'All members',
      CHANNEL_SETTING__MEMBERS__INVITE_MEMBER: 'Invite users',
      CHANNEL_SETTING__MEMBERS__YOU: ' (You)',
      CHANNEL_SETTING__LEAVE_CHANNEL__TITLE: 'Leave channel',
      CHANNEL_SETTING__OPERATORS__TITLE: 'Operators',
      CHANNEL_SETTING__OPERATORS__TITLE_ALL: 'All operators',
      CHANNEL_SETTING__OPERATORS__TITLE_ADD: 'Add operator',
      CHANNEL_SETTING__MUTED_MEMBERS__TITLE: 'Muted members',
      CHANNEL_SETTING__BANNED_MEMBERS__TITLE: 'Banned members',
      CHANNEL_SETTING__FREEZE_CHANNEL: 'Freeze Channel',
      BUTTON__CANCEL: 'Cancel',
      BUTTON__DELETE: 'Delete',
      BUTTON__SAVE: 'Save',
      BUTTON__CREATE: 'Create',
      BUTTON__INVITE: 'Invite',
      BADGE__OVER: '+',
      MODAL__DELETE_MESSAGE__TITLE: 'Delete this message?',
      MODAL__CHANNEL_INFORMATION__TITLE: 'Edit channel information',
      MODAL__CHANNEL_INFORMATION__CHANNEL_IMAGE: 'Channel image',
      MODAL__CHANNEL_INFORMATION__UPLOAD: 'Upload',
      MODAL__CHANNEL_INFORMATION__CHANNEL_NAME: 'Channel name',
      MODAL__CHANNEL_INFORMATION__INPUT__PLACE_HOLDER: 'Enter name',
      MODAL__INVITE_MEMBER__TITLE: 'Invite member',
      MODAL__INVITE_MEMBER__SELECTEC: 'selected',
      MODAL__CHOOSE_CHANNEL_TYPE__TITLE: 'New channel',
      MODAL__CHOOSE_CHANNEL_TYPE__GROUP: 'Group',
      MODAL__CHOOSE_CHANNEL_TYPE__SUPER_GROUP: 'Super group',
      MODAL__CHOOSE_CHANNEL_TYPE__BROADCAST: 'Broadcast',
      MODAL__CREATE_CHANNEL__TITLE: 'New channel',
      MODAL__CREATE_CHANNEL__SELECTED: 'selected',
      TYPING_INDICATOR__IS_TYPING: 'is typing...',
      TYPING_INDICATOR__AND: 'and',
      TYPING_INDICATOR__ARE_TYPING: 'are typing...',
      TYPING_INDICATOR__MULTIPLE_TYPING: 'Several people are typing...',
      CHANNEL_FROZEN: 'Channel frozen',
      PLACE_HOLDER__NO_CHANNEL: 'No channels',
      PLACE_HOLDER__WRONG: 'Something went wrong',
      PLACE_HOLDER__RETRY_TO_CONNECT: 'Retry',
      PLACE_HOLDER__NO_MESSAGES: 'No messages',
      NO_TITLE: 'No title',
      NO_NAME: '(No name)',
      NO_MEMBERS: '(No members)',
      TOOLTIP__AND_YOU: ', and you',
      TOOLTIP__YOU: 'you',
      TOOLTIP__UNKNOWN_USER: '(no name)',
      UNKNOWN__UNKNOWN_MESSAGE_TYPE: '(Unknown message type)',
      UNKNOWN__CANNOT_READ_MESSAGE: 'Cannot read this message.',
      MESSAGE_EDITED: '(edited)',
      MESSAGE_MENU__COPY: 'Copy',
      MESSAGE_MENU__REPLY: 'Reply',
      MESSAGE_MENU__EDIT: 'Edit',
      MESSAGE_MENU__RESEND: 'Resend',
      MESSAGE_MENU__DELETE: 'Delete',
      SEARCH: 'Search',
      SEARCH_IN_CHANNEL: 'Search in channel',
      SEARCH_IN: 'Search in',
      SEARCHING: 'Searching for messages...',
      NO_SEARCHED_MESSAGE: 'No results found.',
      REPLYING_ATTACHMENT__REPLY_TO: 'Reply to',
      REPLYING_ATTACHMENT__FILE_TYPE__IMAGE: 'Photo',
      REPLYING_ATTACHMENT__FILE_TYPE__GIF: 'GIF',
      REPLYING_ATTACHMENT__FILE_TYPE__VIDEO: 'Video',
      REPLIED_TO: 'replied to',
      // FIXME: get back legacy, remove after refactoring open channel messages
      CONTEXT_MENU_DROPDOWN__COPY: 'Copy',
      CONTEXT_MENU_DROPDOWN__EDIT: 'Edit',
      CONTEXT_MENU_DROPDOWN__RESEND: 'Resend',
      CONTEXT_MENU_DROPDOWN__DELETE: 'Delete',
    },
  };
  return stringSet[lang];
};

export default getStringSet;
