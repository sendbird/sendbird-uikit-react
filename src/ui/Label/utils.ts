import { Typography, Colors } from './types';
import { ObjectValues } from '../../utils/typeHelpers/objectValues';

export function changeTypographyToClassName(type?: ObjectValues<typeof Typography>) {
  switch (type) {
    case Typography.H_1: return 'sendbird-label--h-1';
    case Typography.H_2: return 'sendbird-label--h-2';
    case Typography.SUBTITLE_1: return 'sendbird-label--subtitle-1';
    case Typography.SUBTITLE_2: return 'sendbird-label--subtitle-2';
    case Typography.BODY_1: return 'sendbird-label--body-1';
    case Typography.BODY_2: return 'sendbird-label--body-2';
    case Typography.BUTTON_1: return 'sendbird-label--button-1';
    case Typography.BUTTON_2: return 'sendbird-label--button-2';
    case Typography.BUTTON_3: return 'sendbird-label--button-3';
    case Typography.CAPTION_1: return 'sendbird-label--caption-1';
    case Typography.CAPTION_2: return 'sendbird-label--caption-2';
    case Typography.CAPTION_3: return 'sendbird-label--caption-3';
    default: return '';
  }
}

export function changeColorToClassName(color?: ObjectValues<typeof Colors>) {
  switch (color) {
    case Colors.ONBACKGROUND_1: return 'sendbird-label--color-onbackground-1';
    case Colors.ONBACKGROUND_2: return 'sendbird-label--color-onbackground-2';
    case Colors.ONBACKGROUND_3: return 'sendbird-label--color-onbackground-3';
    case Colors.ONBACKGROUND_4: return 'sendbird-label--color-onbackground-4';
    case Colors.ONCONTENT_1: return 'sendbird-label--color-oncontent-1';
    case Colors.ONCONTENT_2: return 'sendbird-label--color-oncontent-2';
    case Colors.PRIMARY: return 'sendbird-label--color-primary'; // should be Primary-3 fix me
    case Colors.ERROR: return 'sendbird-label--color-error';
    case Colors.SECONDARY_3: return 'sendbird-label--color-secondary-3';
    default: return '';
  }
}
