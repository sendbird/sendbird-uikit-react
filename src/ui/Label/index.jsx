import React from 'react';
import PropTypes from 'prop-types';

import './index.scss';
import { Typography, Colors } from './types';
import { changeTypographyToClassName, changeColorToClassName } from './utils';
import getStringSet from './stringSet';

export default function Label({
  className,
  type,
  color,
  children,
}) {
  return (
    <span
      className={[
        ...(Array.isArray(className) ? className : [className]),
        'sendbird-label',
        type ? changeTypographyToClassName(type) : '',
        color ? changeColorToClassName(color) : '',
      ].join(' ')}
    >
      {children}
    </span>
  );
}

Label.propTypes = {
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string),
  ]),
  type: PropTypes.oneOf([
    'H_1',
    'H_2',
    'SUBTITLE_1',
    'SUBTITLE_2',
    'BODY_1',
    'BODY_2',
    'BUTTON_1',
    'BUTTON_2',
    'CAPTION_1',
    'CAPTION_2',
    'CAPTION_3',
  ]),
  color: PropTypes.oneOf([
    'ONBACKGROUND_1',
    'ONBACKGROUND_2',
    'ONBACKGROUND_3',
    'ONBACKGROUND_4',
    'ONCONTENT_1',
    'ONCONTENT_2',
    'PRIMARY',
    'ERROR',
    'SECONDARY_3',
  ]),
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.element,
    PropTypes.any,
  ]),
};

Label.defaultProps = {
  className: [],
  type: '',
  color: '',
  children: null,
};

const LabelTypography = Typography;
const LabelColors = Colors;
const LabelStringSet = getStringSet('en');
export { LabelTypography, LabelColors, LabelStringSet };
