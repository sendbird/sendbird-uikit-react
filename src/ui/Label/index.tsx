import React, {RefObject} from 'react';

import './index.scss';
import { Typography, Colors } from './types';
import { changeTypographyToClassName, changeColorToClassName } from './utils';
import getStringSet from './stringSet';
import { ObjectValues } from '../../utils/typeHelpers/objectValues';
import {ThreadReplies} from '../ThreadReplies';

type LabelProps = {
  className?: string | string[];
  type?: ObjectValues<typeof Typography>;
  color?: ObjectValues<typeof Colors>;
  children?: React.ReactNode;
};
export function Label(
  {
    className = [],
    type,
    color,
    children = null,
  }: LabelProps,
  ref?: RefObject<HTMLDivElement>,
) {
  return (
    // Donot make this into div
    // Mention uses Label. If we use div, it would break the mention detection on Paste
    // https://github.com/sendbird/sendbird-uikit-react/pull/479
    <span
      className={[
        ...(Array.isArray(className) ? className : [className]),
        'sendbird-label',
        changeTypographyToClassName(type),
        changeColorToClassName(color),
      ].join(' ')}
      ref={ref}
    >
      {children}
    </span>
  );
}

const LabelTypography = Typography;
const LabelColors = Colors;
const LabelStringSet = getStringSet('en');
export { LabelTypography, LabelColors, LabelStringSet };

export default React.forwardRef(Label);