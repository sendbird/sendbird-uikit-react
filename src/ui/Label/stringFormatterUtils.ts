import {CoreMessageType, isMultipleFilesMessage} from "../../utils";
import {match} from "ts-pattern";
import {MultipleFilesMessage} from "@sendbird/chat/message";

export function getModalDeleteMessageTitle(stringSet: Record<string, string>, message: CoreMessageType): string {
  return match(message)
    .when(isMultipleFilesMessage, () => {
      const filesCount: number = (message as MultipleFilesMessage).fileInfoList.length;
      return `Do you want to delete all ${filesCount} photos?`;
    })
    .otherwise(() => {
      return stringSet.MODAL__DELETE_MESSAGE__TITLE;
    });
}
