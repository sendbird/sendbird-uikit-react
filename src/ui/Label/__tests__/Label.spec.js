import React from 'react';
import { render, screen } from '@testing-library/react';

import Label, { LabelTypography, LabelColors } from '../index.jsx';
import { changeTypographyToClassName, changeColorToClassName } from '../utils.js';

describe('ui/Label', () => {
  it('should have default classname if props are not provided', function () {
    const text = 'Example';
    const { container } = render(
      <Label
        type={LabelTypography.H_1}
        color={LabelColors.ONBACKGROUND_1}
      >
        {text}
      </Label>
    );
    expect(
      container.querySelector('.sendbird-label').className
    ).toContain('sendbird-label--h-1');
    expect(
      container.querySelector('.sendbird-label').className
    ).toContain('sendbird-label--color-onbackground-1');
    expect(
      screen.getByText(text).className
    ).toContain('sendbird-label');
  });

  it('should have each typography', function () {
    for (let color in LabelColors) {
      for (let typography in LabelTypography) {
        const { container } = render(<Label color={color} type={typography}>Component</Label>);
        expect(
          container.querySelector('.sendbird-label').className
        ).toContain(changeColorToClassName(color));
        expect(
          container.querySelector('.sendbird-label').className
        ).toContain(changeTypographyToClassName(typography));
      }
    }
  });

  it('should create a snapshot of a default label component', function () {
    const { asFragment } = render(
      <Label
        type={LabelTypography.H_1}
        color={LabelColors.ONBACKGROUND_1}
      >
        Example
      </Label>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
