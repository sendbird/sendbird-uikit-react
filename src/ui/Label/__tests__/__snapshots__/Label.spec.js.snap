// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/Label should create a snapshot of a default label component 1`] = `
<DocumentFragment>
  <span
    class="sendbird-label sendbird-label--h-1 sendbird-label--color-onbackground-1"
  >
    Example
  </span>
</DocumentFragment>
`;
